File: Leetcode/1254 Number of Closed Islands.py
VISITED_LAND = -1

class Solution:
    def closedIsland(self, grid):
        n, m = len(grid), len(grid[0])
        def dfs(i, j):
            if i < 0 or i == n or j < 0 or j == m:
                return 0
            if grid[i][j] != 0:
                return 1
            grid[i][j] = VISITED_LAND
            return min(dfs(i-1, j), dfs(i, j+1), dfs(i+1, j), dfs(i, j-1))
        
        n_closed = 0
        for i in range(1, n-1):
            for j in range(1, m-1):
                if grid[i][j] == 0:
                    n_closed += dfs(i, j)
        return n_closed

File: Leetcode/1254 Number of Closed Islands.ts
const ON_EDGE = -1, NOT_ON_EDGE = 1, LAND = 0, VISITED_LAND = 2;

function closedIsland(grid: number[][]): number {
    const n = grid.length, m = grid[0].length;
    let nClosedIslands = 0;
    for(let i = 1; i < (n-1); i++) {
        for(let j = 1; j < (m-1); j++) {
            if(grid[i][j] !== LAND) {
                continue;
            }
            if(dfs(i, j, grid) === NOT_ON_EDGE) {
                nClosedIslands++;
            }
        }
    }
    return nClosedIslands;
}

function dfs(i: number, j: number, grid: number[][]) {
    const n = grid.length, m = grid[0].length;
    if(i < 0 || i >= n || j < 0 || j >= m) {
        return ON_EDGE;
    }
    if(grid[i][j] !== 0) {
        return NOT_ON_EDGE;
    }
    grid[i][j] = VISITED_LAND;
    const neighbourStates = [
        dfs(i - 1, j, grid), 
        dfs(i ,j + 1, grid), 
        dfs(i + 1, j, grid), 
        dfs(i, j - 1, grid)
    ];
    return neighbourStates.some(state => state === ON_EDGE) ? ON_EDGE : NOT_ON_EDGE;
}
